tools/pinning/DESIGN.md

```diff
@@ -0,0 +1,124 @@
+present a problem for us pinning dependencies because they offer a well defined
+format where the dependencies of a package can be easily parsed.
+
+Source distributions or "sdists" are a problem though because they can contain
+arbitrary Python code that must be run to determine the dependencies of the
+package. This code could theoretically do anything, but it most commonly
+inspects the environment it is running in and changes the project's
+dependencies based the environment. As of writing this, we even do this in some
+of Certbot's packages as you can see
+[here](https://github.com/certbot/certbot/blob/8b610239bfcf7aac06f6e36d09a5abba3ba87047/certbot-dns-cloudflare/setup.py#L15-L27).
+This is a problem because it means the environment an sdist was run in affects
+the dependencies it declares making it difficult for us to determine Certbot's
+dependencies for an arbitrary environment. Luckily, this is becoming less and
+less of an issue with the increasing use of wheels, however, as of writing
+this, some of Certbot's dependencies are still only available as sdists on some
+platforms.
+
+### Environment markers and pyproject.toml
+
+Two other things have helped reduce the problems caused by sdists and are
+relevant here. The first is the usage of [environment
+markers](https://www.python.org/dev/peps/pep-0496/) which allows packages to
+consistently declare its conditional dependencies with a static string
+specifying the conditions where a dependency is required instead of dynamically
+generating the list of required dependencies at runtime. This static string
+keeps the package's declaration of its dependencies consistent across
+environments.
+
+The other relatively recent change in Python packaging is the adoption of
+[pyproject.toml files](https://www.python.org/dev/peps/pep-0518/) which allows
+sdists to define their packages using a static file instead of a setup.py
+file which I expect most Python developers are more familiar with as I'm
+writing this. Using a static file instead of arbitrary Python code makes it
+much easier for package declarations to be reliably interpreted. The
+introduction of pyproject.toml also allows for the use of build systems other
+than setuptools which becomes relevant in the next section of this doc.
+
+## Our pinning system
+
+### Overview
+
+The files under this directory are used to generate Certbot's pinning files.
+The files under `oldest` are used to generate the constraints file used for our
+"oldest" tests while `normal` is used to generate the constraints used
+everywhere else. `common` includes shared files that are used for both sets of
+pinnings.
+
+Under `normal` and `oldest`, there are two files as of writing this. One is a
+pyproject.toml file for use with [Poetry](https://python-poetry.org/) while
+the other is a script that can be run to regenerate pinnings. The
+pyproject.toml file defines a dummy Python package that depends on everything
```

Comment:
You're right that "dummy" is probably a bad word to use here. A valid package is being defined. The package is just kind of unusual because it contains no Python code at all and instead it just contains the (meta)data defined in `pyproject.toml`. It's equivalent to a Python project with only a `setup.py` file.

I went ahead and just removed all uses of the word "dummy" which I don't personally think really loses us anything other than future potential confusion about this word choice.

Reply:
Great, thank you!